topology/probes/k8sctl.go

```diff
@@ -0,0 +1,537 @@
+	nss, err := k8s.clientset.CoreV1().Namespaces().List(v1.ListOptions{})
+	if err != nil {
+		logging.GetLogger().Error(err.Error())
+	}
+
+	var i int
+	for _, ns := range nss.Items {
+
+		var deny = false
+		// get ns annotations
+		if len(ns.Annotations) > 0{
+			for _,v := range ns.Annotations {
+				if strings.Contains(v,"DefaultDeny") {
+					logging.GetLogger().Debugf("\t\t DefaultDeny\n")
+					deny = true
+				}
+			}
+		}
+
+		// get network policies
+		nps, err := k8s.clientset.ExtensionsV1beta1().NetworkPolicies(ns.GetName()).List(v1.ListOptions{})
+		if err != nil {
+			logging.GetLogger().Debugf("Error: "+err.Error())
+		}
+		logging.GetLogger().Debugf("%s ns: There are %d network policies in the cluster\n\n", ns.GetName(),len(nps.Items))
+
+		for _, np := range nps.Items {
+
+			var npText string
+			var FromPod []k8sPodEdge
+			var ToPod []k8sPodEdge
+			for ingress := range np.Spec.Ingress {
+				// Only single descriptor is used to choose the "From" np
+				for from := range np.Spec.Ingress[ingress].From {
+					PodSelectorLabels := np.Spec.Ingress[ingress].From[from].PodSelector
+					if (PodSelectorLabels != nil) {
+
+						for key, value := range  PodSelectorLabels.MatchLabels {
+							id := k8s.findPodByLabel(key,value,"")
+							if id.Valid{
+								FromPod = append(FromPod, id)
+							}
+						}
+						break
+					}
+
+					NamespaceSelectorLabels := np.Spec.Ingress[ingress].From[from].NamespaceSelector
+					if (NamespaceSelectorLabels != nil) {
+						for _, value := range  NamespaceSelectorLabels.MatchLabels {
+							// key/value for ns are actually the namespace name itself
+							valid, id := k8s.findPodByNS(value)
+							if valid{
+								FromPod = id
+							}
+						}
+						break
+					}
+				}
+
+				if len(np.Spec.Ingress[ingress].Ports)>0 {
+
+					for port := range np.Spec.Ingress[ingress].Ports {
+						ProtocolStr := valueToStringGenerated(np.Spec.Ingress[ingress].Ports[port].Protocol)
+						npText = fmt.Sprintf("%s %s/%s ",npText, ProtocolStr, np.Spec.Ingress[ingress].Ports[port].Port)
+					}
+				}
+			}
+
+			if len(np.Spec.PodSelector.MatchLabels)>0 {
+				for key, value := range  np.Spec.PodSelector.MatchLabels {
+					id := k8s.findPodByLabel(key,value,ns.GetName())
+					if id.Valid{
+						ToPod = append(ToPod, id)
+					}
+				}
+			} else {
+				//choose all pods in the given ns
+				valid, pod := k8s.findPodByNS(ns.GetName())
+				if valid{
+					ToPod = pod
+				}
+			}
+
+			// build the np edge - skip if not deny - to avoid system ns which allows all-to-all communication between their pods
+			if deny {
+				for _,fromp := range FromPod {
+					for _,top := range ToPod {
+						logging.GetLogger().Debugf("fromp.GraphNode = %s  top.GraphNode = %s \n", fromp.GraphNode, top.GraphNode)
+						if ((fromp.GraphNode != "") && (top.GraphNode != "")) {
+							var direction = "->"
+							if strings.Compare(npText,"") != 0 {
+								npText = fmt.Sprintf("ingress: %s ",npText)
```

Comment:
do you think we could have something structured here instead of text. does this allow us the do requests ?

Reply:
npText is used for the text on the link. I suggest to add edge metadata in the next versions, for query purposes.

Reply:
actually the network policy edge already contains metadata....